File: apps/demo-be-e2e/tests/steps/auth/token-lifecycle.steps.ts
import { expect } from "@playwright/test";
import { createBdd } from "playwright-bdd";
import { setResponse } from "../../utils/response-store";
import {
  getTokenForUser,
  getRefreshTokenForUser,
  setRefreshTokenForUser,
  setTokenForUser,
} from "../../utils/token-store";

const { Given, When, Then } = createBdd();

// ---------------------------------------------------------------------------
// Token lifecycle steps
// ---------------------------------------------------------------------------

When(/^alice sends POST \/api\/v1\/auth\/refresh with her refresh token$/, async ({ request }) => {
  const token = getRefreshTokenForUser("alice");
  setResponse(
    await request.post("/api/v1/auth/refresh", {
      data: { refreshToken: token },
      headers: { "Content-Type": "application/json" },
    }),
  );
});

Given("alice's refresh token has expired", async () => {
  // The refresh token stored is valid; we simulate expiry by overwriting with
  // an obviously invalid/expired token value so the next refresh call returns 401.
  setRefreshTokenForUser("alice", "expired.refresh.token");
});

Given("alice has used her refresh token to get a new token pair", async ({ request }) => {
  const originalRefresh = getRefreshTokenForUser("alice");
  // Perform a refresh to rotate the token pair
  const res = await request.post("/api/v1/auth/refresh", {
    data: { refreshToken: originalRefresh },
    headers: { "Content-Type": "application/json" },
  });
  const body = (await res.json()) as Record<string, unknown>;
  // Store new tokens but keep the original refresh token accessible via a
  // secondary key so the scenario can still send it
  setTokenForUser("alice_new", body["accessToken"] as string);
  setRefreshTokenForUser("alice_new", body["refreshToken"] as string);
  // alice's refresh token remains as the original (for the "send original" step)
});

When(/^alice sends POST \/api\/v1\/auth\/refresh with her original refresh token$/, async ({ request }) => {
  const token = getRefreshTokenForUser("alice");
  setResponse(
    await request.post("/api/v1/auth/refresh", {
      data: { refreshToken: token },
      headers: { "Content-Type": "application/json" },
    }),
  );
});

Given("the user {string} has been deactivated", async ({ request }, username: string) => {
  // Need an admin to disable this user, or we self-deactivate
  // Since we just need the user deactivated and we have their token:
  const token = getTokenForUser(username);
  await request.post("/api/v1/users/me/deactivate", {
    headers: { Authorization: `Bearer ${token}` },
  });
});

When(/^alice sends POST \/api\/v1\/auth\/logout-all with her access token$/, async ({ request }) => {
  const token = getTokenForUser("alice");
  setResponse(
    await request.post("/api/v1/auth/logout-all", {
      headers: { Authorization: `Bearer ${token}` },
    }),
  );
});

Given("alice has already logged out once", async ({ request }) => {
  const token = getTokenForUser("alice");
  await request.post("/api/v1/auth/logout", {
    headers: { Authorization: `Bearer ${token}` },
  });
});

Then("alice's access token should be invalidated", async ({ request }) => {
  const token = getTokenForUser("alice");
  const res = await request.get("/api/v1/users/me", {
    headers: { Authorization: `Bearer ${token}` },
  });
  expect(res.status()).toBe(401);
});

File: apps/demo-be-python-fastapi/tests/unit/steps/token_lifecycle_steps.py
"""BDD step definitions for token lifecycle feature."""

from pytest_bdd import given, scenarios, then, when

from demo_be_python_fastapi.auth.jwt_service import create_expired_refresh_token
from tests.integration.service_client import FakeResponse, ServiceClient
from tests.unit.conftest import GHERKIN_ROOT

scenarios(str(GHERKIN_ROOT / "authentication" / "token-lifecycle.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    '"alice" has logged in and stored the access token and refresh token',
    target_fixture="alice_tokens",
)
def alice_login_tokens(client: ServiceClient, registered_user: dict) -> dict:
    return client.login_user("alice", _PASSWORD)


@given("alice's refresh token has expired", target_fixture="alice_tokens")
def alice_expired_refresh_token(client: ServiceClient, registered_user: dict) -> dict:
    # Get a real access token but substitute an expired refresh token
    tokens = client.login_user("alice", _PASSWORD)
    tokens["refreshToken"] = create_expired_refresh_token(registered_user["id"])
    return tokens


@given(
    "alice has used her refresh token to get a new token pair",
    target_fixture="alice_tokens",
)
def alice_used_refresh_token(client: ServiceClient, registered_user: dict) -> dict:
    tokens = client.login_user("alice", _PASSWORD)
    original_refresh = tokens["refreshToken"]
    # Use the refresh token once so it becomes revoked
    resp = client.post_refresh(original_refresh)
    assert resp.status_code == 200
    # Return the tokens with the original (now revoked) refresh token stored separately
    return {**tokens, "_original_refresh": original_refresh}


@given('the user "alice" has been deactivated', target_fixture="deactivated_alice")
def deactivate_alice(client: ServiceClient, registered_user: dict, alice_tokens: dict) -> dict:
    resp = client.post_me_deactivate(f"Bearer {alice_tokens['accessToken']}")
    assert resp.status_code == 200
    return registered_user


@given("alice has already logged out once")
def alice_already_logged_out(
    client: ServiceClient, registered_user: dict, alice_tokens: dict
) -> None:
    client.post_logout(f"Bearer {alice_tokens['accessToken']}")


# --- When steps ---


@when("alice sends POST /api/v1/auth/refresh with her refresh token", target_fixture="response")
def alice_refresh(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.post_refresh(alice_tokens["refreshToken"])


_STEP_REFRESH_ORIGINAL = "alice sends POST /api/v1/auth/refresh with her original refresh token"


@when(_STEP_REFRESH_ORIGINAL, target_fixture="response")
def alice_refresh_with_original(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    original = alice_tokens.get("_original_refresh", alice_tokens["refreshToken"])
    return client.post_refresh(original)


@when("alice sends POST /api/v1/auth/logout with her access token", target_fixture="response")
def alice_logout(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.post_logout(f"Bearer {alice_tokens['accessToken']}")


@when("alice sends POST /api/v1/auth/logout-all with her access token", target_fixture="response")
def alice_logout_all(client: ServiceClient, alice_tokens: dict) -> FakeResponse:
    return client.post_logout_all(f"Bearer {alice_tokens['accessToken']}")


# --- Then steps ---


@then("alice's access token should be invalidated")
def check_access_token_invalidated(client: ServiceClient, alice_tokens: dict) -> None:
    resp = client.get_me(f"Bearer {alice_tokens['accessToken']}")
    assert resp.status_code == 401, f"Expected 401 after logout, got {resp.status_code}"
